# WE RATE DOGS PROJECT

In [1]:
import pandas as pd
import numpy as np
import tweepy
from tweepy import OAuthHandler
import json
from timeit import default_timer as timer

## 1. Gather

### 1.1 Enhanced Twitter Archive Data

In [2]:
archive_df = pd.read_csv('twitter-archive-enhanced.csv')
print('Archive data has: ' + str(len(archive_df)) + ' observations')
archive_df.sample(3)

Archive data has: 2356 observations


tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
834   768473857036525572                    NaN                  NaN   
268   841439858740625411                    NaN                  NaN   
1982  672968025906282496                    NaN                  NaN   

                      timestamp  \
834   2016-08-24 15:43:39 +0000   
268   2017-03-14 00:04:30 +0000   
1982  2015-12-05 02:37:35 +0000   

                                                 source  \
834   <a href="http://twitter.com/download/iphone" r...   
268   <a href="http://twitter.com/download/iphone" r...   
1982  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
834   Meet Chevy. He had a late breakfast and now ha...                  NaN   
268   Here we have some incredible doggos for #K9Vet...                  NaN   
1982  This is Gustaf. He's a purebred Chevy Equinox....                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
834                        NaN                        NaN   
268                        NaN                        NaN   
1982                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
834   https://twitter.com/dog_rates/status/768473857...                11   
268   https://twitter.com/dog_rates/status/841439858...                14   
1982  https://twitter.com/dog_rates/status/672968025...                11   

      rating_denominator    name doggo floofer pupper puppo  
834                   10   Chevy  None    None   None  None  
268                   10    None  None    None   None  None  
1982                  10  Gustaf  None    None   None  None

### 1.2 Image Predictions Data

In [3]:
img_pred_df = pd.read_csv('image-predictions.tsv', sep='\t')
print('Image Predictions data has: ' + str(len(img_pred_df)) + ' observations')
img_pred_df.sample(3)

Image Predictions data has: 2075 observations


tweet_id                                          jpg_url  \
346   672466075045466113  https://pbs.twimg.com/media/CVUUU_EWoAAxABV.jpg   
1025  710658690886586372  https://pbs.twimg.com/media/CdzETn4W4AAVU5N.jpg   
1859  842115215311396866  https://pbs.twimg.com/media/C6_LTCZWoAAKm_O.jpg   

      img_num                           p1   p1_conf  p1_dog              p2  \
346         1               cocker_spaniel  0.150424    True      toy_poodle   
1025        1  soft-coated_wheaten_terrier  0.948617    True  Dandie_Dinmont   
1859        1                         chow  0.293493    True    Newfoundland   

       p2_conf  p2_dog                      p3   p3_conf  p3_dog  
346   0.088605    True  Welsh_springer_spaniel  0.072014    True  
1025  0.018664    True                   cairn  0.015943    True  
1859  0.181336    True              schipperke  0.125152    True

### 1.3 Twitter Scraped Data

In [4]:
# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
# These are hidden to comply with Twitter's API terms and conditions
consumer_key = 'hRwnPAXiqvoWWvBdndaQfVgVC'
consumer_secret = '3KyvKdER17GOvY7coL5pZGG0n4iZqXyLAJeS240K8QXWQo1OHd'
access_token = '1204094449363673094-Bw8332TZG11wCRfB8L7XzkUUUASBKL'
access_secret = 'bHL5BEgFeBXuMlOp4GL0J9bs3GdHmPHyGAIhgqku8ZJaI'

auth = OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True)

In [5]:
'''# NOTE TO STUDENT WITH MOBILE VERIFICATION ISSUES:
# df_1 is a DataFrame with the twitter_archive_enhanced.csv file. You may have to
# change line 17 to match the name of your DataFrame with twitter_archive_enhanced.csv
# NOTE TO REVIEWER: this student had mobile verification issues so the following
# Twitter API code was sent to this student from a Udacity instructor
# Tweet IDs for which to gather additional data via Twitter's API
tweet_ids = archive_df.tweet_id.values
#len(tweet_ids)

# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
count = 0
fails_dict = {}
start = timer()
# Save each tweet's returned JSON as a new line in a .txt file
with open('tweet_json.txt', 'w') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
    for tweet_id in tweet_ids:
        #count += 1
        #print(str(count) + ": " + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            json.dump(tweet._json, outfile)
            outfile.write('\n')
        except tweepy.TweepError as e:
            fails_dict[tweet_id] = e
            pass
end = timer()
print("total time: " + str(end - start))
print("The following tweet ids have failed:")
print(fails_dict)
'''

'# NOTE TO STUDENT WITH MOBILE VERIFICATION ISSUES:\n# df_1 is a DataFrame with the twitter_archive_enhanced.csv file. You may have to\n# change line 17 to match the name of your DataFrame with twitter_archive_enhanced.csv\n# NOTE TO REVIEWER: this student had mobile verification issues so the following\n# Twitter API code was sent to this student from a Udacity instructor\n# Tweet IDs for which to gather additional data via Twitter\'s API\ntweet_ids = archive_df.tweet_id.values\n#len(tweet_ids)\n\n# Query Twitter\'s API for JSON data for each tweet ID in the Twitter archive\ncount = 0\nfails_dict = {}\nstart = timer()\n# Save each tweet\'s returned JSON as a new line in a .txt file\nwith open(\'tweet_json.txt\', \'w\') as outfile:\n    # This loop will likely take 20-30 minutes to run because of Twitter\'s rate limit\n    for tweet_id in tweet_ids:\n        #count += 1\n        #print(str(count) + ": " + str(tweet_id))\n        try:\n            tweet = api.get_status(tweet_id, tweet_

In [6]:
tweets = []
for line in open('tweet_json.txt', 'r'):
    tweets.append(json.loads(line))
scraped_df = pd.DataFrame(tweets)
print('Scraped data has: ' + str(len(scraped_df)) + ' observations')
scraped_df.sample(3)

Scraped data has: 2326 observations


created_at                  id              id_str  \
866  Sat Jul 30 01:22:17 +0000 2016  759197388317847553  759197388317847553   
617  Tue Nov 01 02:00:14 +0000 2016  793271401113350145  793271401113350145   
670  Sat Oct 15 16:01:13 +0000 2016  787322443945877504  787322443945877504   

                                             full_text  truncated  \
866  This is Luna. She's just heckin precious af I ...      False   
617  Meet Dave. It's his favorite day of the year. ...      False   
670  This is Lincoln. He forgot to use his blinker ...      False   

    display_text_range                                           entities  \
866            [0, 77]  {'hashtags': [], 'symbols': [], 'user_mentions...   
617           [0, 118]  {'hashtags': [], 'symbols': [], 'user_mentions...   
670           [0, 105]  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                     extended_entities  \
866  {'media': [{'id': 759197380596170757, 'id_str'...   
617  {'media': [{'id': 793271391365783552, 'id_str'...   
670  {'media': [{'id': 787322437922873345, 'id_str'...   

                                                source  in_reply_to_status_id  \
866  <a href="http://twitter.com/download/iphone" r...                    NaN   
617  <a href="http://twitter.com/download/iphone" r...                    NaN   
670  <a href="http://twitter.com/download/iphone" r...                    NaN   

     ... favorited  retweeted possibly_sensitive  \
866  ...     False      False              False   
617  ...     False      False              False   
670  ...     False      False              False   

    possibly_sensitive_appealable lang retweeted_status quoted_status_id  \
866                         False   en              NaN              NaN   
617                         False   en              NaN              NaN   
670                         False   en              NaN              NaN   

    quoted_status_id_str quoted_status_permalink  quoted_status  
866                  NaN                     NaN            NaN  
617                  NaN                     NaN            NaN  
670                  NaN                     NaN            NaN  

[3 rows x 32 columns]

## Assessment

We can clearly see immediately that we were not able to retrieve the data for 30 tweet_ids from twitter scraping, hence only 2326 tweets. Also image predictions also has only 2074 tweets. This causes completeness issue in our data.

In [7]:
archive_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

We notice that in_reply_status_id and retweeted_status_id are non-null for certain rows. Implyig those rows are either replies/retweets. Those are relevant data to our project.

There is also 4 different columns for doggo/floofer/pupper/puppo which could one column

In [8]:
scraped_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2326 entries, 0 to 2325
Data columns (total 32 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   created_at                     2326 non-null   object 
 1   id                             2326 non-null   int64  
 2   id_str                         2326 non-null   object 
 3   full_text                      2326 non-null   object 
 4   truncated                      2326 non-null   bool   
 5   display_text_range             2326 non-null   object 
 6   entities                       2326 non-null   object 
 7   extended_entities              2056 non-null   object 
 8   source                         2326 non-null   object 
 9   in_reply_to_status_id          77 non-null     float64
 10  in_reply_to_status_id_str      77 non-null     object 
 11  in_reply_to_user_id            77 non-null     float64
 12  in_reply_to_user_id_str        77 non-null     o

Similarly, we need to remove the from replies/retweets from the scraped data as well

In [9]:
img_pred_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [10]:
archive_df[archive_df.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

In [27]:
img_pred_df[img_pred_df.tweet_id.duplicated()]
img_pred_df[img_pred_df.jpg_url.duplicated()]

tweet_id                                            jpg_url  \
1297  752309394570878976  https://pbs.twimg.com/ext_tw_video_thumb/67535...   
1315  754874841593970688    https://pbs.twimg.com/media/CWza7kpWcAAdYLc.jpg   
1333  757729163776290825    https://pbs.twimg.com/media/CWyD2HGUYAQ1Xa7.jpg   
1345  759159934323924993    https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg   
1349  759566828574212096    https://pbs.twimg.com/media/CkNjahBXAAQ2kWo.jpg   
...                  ...                                                ...   
1903  851953902622658560    https://pbs.twimg.com/media/C4KHj-nWQAA3poV.jpg   
1944  861769973181624320    https://pbs.twimg.com/media/CzG425nWgAAnP7P.jpg   
1992  873697596434513921    https://pbs.twimg.com/media/DA7iHL5U0AA1OQo.jpg   
2041  885311592912609280    https://pbs.twimg.com/media/C4bTH6nWMAAX_bJ.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   

      img_num                         p1   p1_conf  p1_dog  \
1297        1                    upright  0.303415   False   
1315        1                        pug  0.272205    True   
1333        2               cash_machine  0.802333   False   
1345        1              Irish_terrier  0.254856    True   
1349        1         Labrador_retriever  0.967397    True   
...       ...                        ...       ...     ...   
1903        1  Staffordshire_bullterrier  0.757547    True   
1944        2              Arabian_camel  0.366248   False   
1992        1                     laptop  0.153718   False   
2041        1         Labrador_retriever  0.908703    True   
2055        2                   Pembroke  0.809197    True   

                                  p2   p2_conf  p2_dog  \
1297                golden_retriever  0.181351    True   
1315                    bull_mastiff  0.251530    True   
1333                      schipperke  0.045519    True   
1345                          briard  0.227716    True   
1349                golden_retriever  0.016641    True   
...                              ...       ...     ...   
1903  American_Staffordshire_terrier  0.149950    True   
1944                     house_finch  0.209852   False   
1992                  French_bulldog  0.099984    True   
2041                       seat_belt  0.057091   False   
2055             Rhodesian_ridgeback  0.054950    True   

                               p3   p3_conf  p3_dog  
1297             Brittany_spaniel  0.162084    True  
1315                   bath_towel  0.116806   False  
1333              German_shepherd  0.023353    True  
1345  soft-coated_wheaten_terrier  0.223263    True  
1349                     ice_bear  0.014858   False  
...                           ...       ...     ...  
1903     Chesapeake_Bay_retriever  0.047523    True  
1944               cocker_spaniel  0.046403    True  
1992                      printer  0.077130   False  
2041                          pug  0.011933    True  
2055                       beagle  0.038915    True  

[66 rows x 12 columns]

It appears that 66 tweet_ids in image predictions data have duplicated jpg_url indicating that these tweet_ids don't give any extra information hence irrelevant

In [18]:
scraped_df[scraped_df.id.duplicated()]
scraped_df[scraped_df.full_text.duplicated()]

Empty DataFrame
Columns: [created_at, id, id_str, full_text, truncated, display_text_range, entities, extended_entities, source, in_reply_to_status_id, in_reply_to_status_id_str, in_reply_to_user_id, in_reply_to_user_id_str, in_reply_to_screen_name, user, geo, coordinates, place, contributors, is_quote_status, retweet_count, favorite_count, favorited, retweeted, possibly_sensitive, possibly_sensitive_appealable, lang, retweeted_status, quoted_status_id, quoted_status_id_str, quoted_status_permalink, quoted_status]
Index: []

[0 rows x 32 columns]


Empty DataFrame
Columns: [created_at, id, id_str, full_text, truncated, display_text_range, entities, extended_entities, source, in_reply_to_status_id, in_reply_to_status_id_str, in_reply_to_user_id, in_reply_to_user_id_str, in_reply_to_screen_name, user, geo, coordinates, place, contributors, is_quote_status, retweet_count, favorite_count, favorited, retweeted, possibly_sensitive, possibly_sensitive_appealable, lang, retweeted_status, quoted_status_id, quoted_status_id_str, quoted_status_permalink, quoted_status]
Index: []

[0 rows x 32 columns]

In [13]:
scraped_df.describe()

id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.326000e+03           7.700000e+01         7.700000e+01   
mean   7.417666e+17           7.440692e+17         2.040329e+16   
std    6.821069e+16           7.524295e+16         1.260797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.780814e+17           6.757073e+17         3.589728e+08   
50%    7.178159e+17           7.032559e+17         4.196984e+09   
75%    7.986402e+17           8.233264e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweet_count  favorite_count  quoted_status_id  
count    2326.000000     2326.000000      2.600000e+01  
mean     2483.896819     7072.468186      8.113972e+17  
std      4206.941888    10994.571048      6.295843e+16  
min         1.000000        0.000000      6.721083e+17  
25%       497.500000     1225.500000      7.761338e+17  
50%      1154.500000     3055.000000      8.281173e+17  
75%      2876.500000     8614.000000      8.637581e+17  
max     71092.000000   145297.000000      8.860534e+17

The retweet and favourites count being 0 seem like outliers or invalid

In [14]:
archive_df.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

The rating_numerator being 0 seems suspicious. And the rating_denominator is supposed to be always 10. These need further inspection.

In [15]:
img_pred_df.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

The image prediction numerical values look okay.

### Futher Inspection of columns
#### Inspecting duplicates in jpg_url
Taking sample  urls from the list of duplicates and visually inspecting those rows:

In [25]:
img_pred_df[img_pred_df.jpg_url.duplicated()]
#img_pred_df[img_pred_df.jpg_url == 'https://pbs.twimg.com/media/CWza7kpWcAAdYLc.jpg']

tweet_id                                            jpg_url  \
1297  752309394570878976  https://pbs.twimg.com/ext_tw_video_thumb/67535...   
1315  754874841593970688    https://pbs.twimg.com/media/CWza7kpWcAAdYLc.jpg   
1333  757729163776290825    https://pbs.twimg.com/media/CWyD2HGUYAQ1Xa7.jpg   
1345  759159934323924993    https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg   
1349  759566828574212096    https://pbs.twimg.com/media/CkNjahBXAAQ2kWo.jpg   
...                  ...                                                ...   
1903  851953902622658560    https://pbs.twimg.com/media/C4KHj-nWQAA3poV.jpg   
1944  861769973181624320    https://pbs.twimg.com/media/CzG425nWgAAnP7P.jpg   
1992  873697596434513921    https://pbs.twimg.com/media/DA7iHL5U0AA1OQo.jpg   
2041  885311592912609280    https://pbs.twimg.com/media/C4bTH6nWMAAX_bJ.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   

      img_num                         p1   p1_conf  p1_dog  \
1297        1                    upright  0.303415   False   
1315        1                        pug  0.272205    True   
1333        2               cash_machine  0.802333   False   
1345        1              Irish_terrier  0.254856    True   
1349        1         Labrador_retriever  0.967397    True   
...       ...                        ...       ...     ...   
1903        1  Staffordshire_bullterrier  0.757547    True   
1944        2              Arabian_camel  0.366248   False   
1992        1                     laptop  0.153718   False   
2041        1         Labrador_retriever  0.908703    True   
2055        2                   Pembroke  0.809197    True   

                                  p2   p2_conf  p2_dog  \
1297                golden_retriever  0.181351    True   
1315                    bull_mastiff  0.251530    True   
1333                      schipperke  0.045519    True   
1345                          briard  0.227716    True   
1349                golden_retriever  0.016641    True   
...                              ...       ...     ...   
1903  American_Staffordshire_terrier  0.149950    True   
1944                     house_finch  0.209852   False   
1992                  French_bulldog  0.099984    True   
2041                       seat_belt  0.057091   False   
2055             Rhodesian_ridgeback  0.054950    True   

                               p3   p3_conf  p3_dog  
1297             Brittany_spaniel  0.162084    True  
1315                   bath_towel  0.116806   False  
1333              German_shepherd  0.023353    True  
1345  soft-coated_wheaten_terrier  0.223263    True  
1349                     ice_bear  0.014858   False  
...                           ...       ...     ...  
1903     Chesapeake_Bay_retriever  0.047523    True  
1944               cocker_spaniel  0.046403    True  
1992                      printer  0.077130   False  
2041                          pug  0.011933    True  
2055                       beagle  0.038915    True  

[66 rows x 12 columns]